### PR TITLE
First version for the new DA for Cisco routers with IOS XR firmware

### PR DESCRIPTION
**Pull request by yguubiqube**

### CODE REVIEW

**Review by abr-ubiqube — APPROVED** [review]

File: adapters/cisco_ios_xr/cisco_ios_xr_command.php [hunk 1]
```diff
@@ -0,0 +1,424 @@
+<?php
```
Comment (UBiqube-ydu) [inline adapters/cisco_ios_xr/cisco_ios_xr_command.php:1]: All IBA stuff can be removed

File: adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  $SD = &$network->SD;
+
+  $ret = save_result_file($configuration, "conf.applied");
+  if ($ret !== SMS_OK)
+  {
+    return $ret;
+  }
+
+  $SMS_OUTPUT_BUF = '';
+  $line_config_mode = $SD->SD_CONFIG_STEP;
```
Comment (UBiqube-ydu) [inline adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php:37]: lines 37 and 38 are useless

File: adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+  $protocol = $sms_sd_ctx->getParam('PROTOCOL');
+  $ret = SMS_OK;
+
+  $file_name = "{$sdid}.cfg";
```
Comment (UBiqube-ydu) [inline adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php:41]: lines 41 to 49 are useless

File: adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+  $full_name = $_SERVER['TFTP_BASE'] . "/" . $file_name;
+
+  $ret = save_file($configuration, $full_name);
+  if ($ret !== SMS_OK)
+  {
+    sms_log_error(__FILE__ . ':' . __LINE__ . ":save_file Error $ret\n");
+    return $ret;
+  }
+
+/*
```
Comment (UBiqube-ydu) [inline adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php:51]: the entire comment can be removed

File: adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+      sms_log_error(__FILE__ . ':' . __LINE__ . ":SCP Error $ret\n");
+    }
+  }
+*/
+
+  // ---------------------------------------------------
+  // Line by line mode configuration
+  // ---------------------------------------------------
+  $ret = SMS_OK;
+  // if ($line_config_mode === 1)
```
Comment (UBiqube-ydu) [inline adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php:115]: the comment can be removed

File: adapters/cisco_ios_xr/cisco_ios_xr_restore_configuration.php [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		} catch ( Exception | Error $e ) {
+			if (strpos ( $e->getMessage (), 'connection failed' ) !== false) {
+				return ERR_SD_CONNREFUSED;
+			}
+			sms_log_error ( __FILE__ . ':' . __LINE__ . ":SCP Error $ret\n" );
+			// if no tftp attempt, we return
+			return ERR_SD_TFTP;
+		}
+
+		/*
```
Comment (UBiqube-ydu) [inline adapters/cisco_ios_xr/cisco_ios_xr_restore_configuration.php:155]: the comment can be removed

File: adapters/cisco_ios_xr/common.php [hunk 7]
```diff
@@ -0,0 +1,801 @@
...
+      throw new SmsException("Rommon mode after reloading", ERR_SD_FAILED);
+    }
+    else
+    {
+      $end = true;
+    }
+  } while (!$end);
+}
+
+function func_write()
```
Comment (UBiqube-ydu) [inline adapters/cisco_ios_xr/common.php:727]: This function is no more used

File: adapters/cisco_ios_xr/cisco_ios_xr_command.php (hunk 1 above)
Comment (yguubiqube) [reply adapters/cisco_ios_xr/cisco_ios_xr_command.php:1]: This is fixed in commit eb2d0a2

File: adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php (hunk 2 above)
Comment (yguubiqube) [reply adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php:37]: This is fixed in commit 7018ca4

File: adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php (hunk 3 above)
Comment (yguubiqube) [reply adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php:41]: This is fixed in commit 7018ca4

File: adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php (hunk 4 above)
Comment (yguubiqube) [reply adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php:51]: This is fixed in commit 7018ca4

File: adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php (hunk 5 above)
Comment (yguubiqube) [reply adapters/cisco_ios_xr/cisco_ios_xr_apply_conf.php:115]: This is fixed in commit 7018ca4

File: adapters/cisco_ios_xr/cisco_ios_xr_restore_configuration.php (hunk 6 above)
Comment (yguubiqube) [reply adapters/cisco_ios_xr/cisco_ios_xr_restore_configuration.php:155]: This is fixed in commit 7018ca4

File: adapters/cisco_ios_xr/common.php (hunk 7 above)
Comment (yguubiqube) [reply adapters/cisco_ios_xr/common.php:727]: This is fixed in commit ed76f9d

**Pull request merged**